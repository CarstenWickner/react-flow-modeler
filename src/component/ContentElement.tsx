import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export class ContentElement extends React.Component<{
    elementId: string;
    editMenu: (() => React.ReactNode) | undefined;
    onSelect: (elementId: string) => void;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, elementId } = this.props;
        onSelect(elementId);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { editMenu, children } = this.props;
        return (
            <>
                <div className="stroke-horizontal arrow" />
                <div className="flow-element-wrapper">
                    <div className={`flow-element content-element${editMenu ? " selected" : ""}`} onClick={this.onClick}>
                        {children}
                    </div>
                    {editMenu && editMenu()}
                </div>
                <HorizontalStroke optional />
            </>
        );
    }
}
