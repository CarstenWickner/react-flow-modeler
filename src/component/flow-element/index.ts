export { Start } from "./Start";
export { ContentElement } from "./ContentElement";
export { Gateway } from "./Gateway";
export { GatewayConnector } from "./GatewayConnector";
export { StrokeExtension } from "./StrokeExtension";
export { End } from "./End";
