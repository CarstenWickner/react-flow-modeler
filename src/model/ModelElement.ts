export const enum ElementType {
    Start = 1,
    Content,
    GatewayDiverging,
    GatewayConverging,
    ConnectGatewayToElement,
    ConnectElementToGateway,
    End
}

interface BaseModelNode {
    columnIndex: number;
    rowCount: number;
}

export interface StartNode extends BaseModelNode {
    type: ElementType.Start;
    followingElement: ContentNode | DivergingGatewayNode | EndNode;
}

export interface ContentNode extends BaseModelNode {
    type: ElementType.Content;
    id: string;
    precedingElement: StartNode | ContentNode | DivergingGatewayBranch | ConvergingGatewayNode;
    followingElement: ContentNode | DivergingGatewayNode | ConvergingGatewayBranch | EndNode;
    data?: { [key: string]: unknown };
}

export interface DivergingGatewayNode extends BaseModelNode {
    type: ElementType.GatewayDiverging;
    id: string;
    precedingElement: StartNode | ContentNode | DivergingGatewayBranch | ConvergingGatewayNode;
    followingBranches: Array<DivergingGatewayBranch>;
    data?: { [key: string]: unknown };
}

export interface DivergingGatewayBranch extends BaseModelNode {
    type: ElementType.ConnectGatewayToElement;
    precedingElement: DivergingGatewayNode;
    branchIndex: number;
    followingElement: ContentNode | DivergingGatewayNode | ConvergingGatewayBranch;
    data?: { [key: string]: unknown };
}

export interface ConvergingGatewayNode extends BaseModelNode {
    type: ElementType.GatewayConverging;
    precedingBranches: Array<ConvergingGatewayBranch>;
    followingElement: ContentNode | DivergingGatewayNode | EndNode;
}

export interface ConvergingGatewayBranch extends BaseModelNode {
    type: ElementType.ConnectElementToGateway;
    precedingElement: ContentNode | DivergingGatewayBranch;
    followingElement: ConvergingGatewayNode;
    branchIndex: number;
}

export interface EndNode extends BaseModelNode {
    type: ElementType.End;
    precedingElement: StartNode | ContentNode | ConvergingGatewayNode;
}

export type ModelElementExclStart =
    | ContentNode
    | DivergingGatewayNode
    | DivergingGatewayBranch
    | ConvergingGatewayBranch
    | ConvergingGatewayNode
    | EndNode;

export type ModelElement = StartNode | ModelElementExclStart;
