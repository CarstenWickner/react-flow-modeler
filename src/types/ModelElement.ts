export const enum ElementType {
    StartNode = "start",
    StepNode = "step",
    DivergingGatewayNode = "div-gw",
    ConvergingGatewayNode = "conv-gw",
    DivergingGatewayBranch = "div-branch",
    ConvergingGatewayBranch = "conv-branch",
    EndNode = "end"
}

interface BaseModelNode {
    columnIndex: number;
    rowCount: number;
}

export interface StartNode extends BaseModelNode {
    type: ElementType.StartNode;
    followingElement: StepNode | DivergingGatewayNode | EndNode;
}

export interface StepNode extends BaseModelNode {
    type: ElementType.StepNode;
    id: string;
    precedingElement: StartNode | StepNode | DivergingGatewayBranch | ConvergingGatewayNode;
    followingElement: StepNode | DivergingGatewayNode | ConvergingGatewayBranch | EndNode;
    data?: { [key: string]: unknown };
}

export interface DivergingGatewayNode extends BaseModelNode {
    type: ElementType.DivergingGatewayNode;
    id: string;
    precedingElement: StartNode | StepNode | DivergingGatewayBranch | ConvergingGatewayNode;
    followingBranches: Array<DivergingGatewayBranch>;
    data?: { [key: string]: unknown };
}

export interface DivergingGatewayBranch extends BaseModelNode {
    type: ElementType.DivergingGatewayBranch;
    precedingElement: DivergingGatewayNode;
    branchIndex: number;
    followingElement: StepNode | DivergingGatewayNode | ConvergingGatewayBranch;
    data?: { [key: string]: unknown };
}

export interface ConvergingGatewayNode extends BaseModelNode {
    type: ElementType.ConvergingGatewayNode;
    precedingBranches: Array<ConvergingGatewayBranch>;
    followingElement: StepNode | DivergingGatewayNode | EndNode;
}

export interface ConvergingGatewayBranch extends BaseModelNode {
    type: ElementType.ConvergingGatewayBranch;
    precedingElement: StepNode | DivergingGatewayBranch;
    followingElement: ConvergingGatewayNode;
    branchIndex: number;
}

export interface EndNode extends BaseModelNode {
    type: ElementType.EndNode;
    precedingElement: StartNode | StepNode | ConvergingGatewayNode;
}

export type ModelElementExclStart =
    | StepNode
    | DivergingGatewayNode
    | DivergingGatewayBranch
    | ConvergingGatewayBranch
    | ConvergingGatewayNode
    | EndNode;

export type ModelElement = StartNode | ModelElementExclStart;
