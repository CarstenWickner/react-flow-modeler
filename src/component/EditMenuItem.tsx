import * as React from "react";
import { useDrag } from "react-dnd";

import { DraggedLinkContext } from "../types/EditAction";

const disabledDragging: [{ isDragging: boolean }, React.LegacyRef<HTMLSpanElement>] = [{ isDragging: false }, undefined];

export const EditMenuItem: React.FC<{
    options: { className?: string; title?: string };
    defaultClassName: string;
    onClick?: (event: React.MouseEvent) => void;
    dragItem?: DraggedLinkContext;
}> = ({ options, defaultClassName, onClick, dragItem }) => {
    if (!dragItem && !onClick) {
        return null;
    }
    const [{ isDragging }, drag] = dragItem
        ? useDrag({
              item: dragItem,
              collect: (monitor) => ({ isDragging: !!monitor.isDragging() })
          })
        : disabledDragging;
    const baseClassName = (options && options.className) || `menu-item ${defaultClassName}`;
    return (
        <span
            className={`${baseClassName}${isDragging ? " dragging" : ""}`}
            {...{ title: options ? options.title : undefined, ref: drag, onClick }}
        />
    );
};
