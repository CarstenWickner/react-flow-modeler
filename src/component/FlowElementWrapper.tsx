import * as React from "react";
import { useDrop } from "react-dnd";

import { isFlowValid } from "../model/pathValidationUtils";

import { ContentNode, ConvergingGatewayNode, DivergingGatewayNode, ElementType, EndNode, ModelElementExclStart } from "../types/ModelElement";
import { DraggableType, DraggedLinkContext, onLinkDropCallback } from "../types/EditAction";

const isTargetAncestorOfDragItem = (
    originElement: ModelElementExclStart,
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode
): boolean => {
    if (originElement === referenceElement) {
        return true;
    }
    if (originElement.type === ElementType.ConvergingGatewayNode) {
        return originElement.precedingBranches.some((branch) => isTargetAncestorOfDragItem(branch, referenceElement));
    }
    return (
        originElement.precedingElement.type !== ElementType.StartNode && isTargetAncestorOfDragItem(originElement.precedingElement, referenceElement)
    );
};

const isDropValid = (referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode, onDrop: onLinkDropCallback) => (
    dragContext: DraggedLinkContext
): boolean =>
    !isTargetAncestorOfDragItem(dragContext.originElement, referenceElement) &&
    dragContext.originElement.followingElement !== referenceElement &&
    isFlowValid(onDrop(referenceElement, dragContext, true).changedFlow);

const disabledDropping: [{ isOver: boolean; canDrop: false }, React.LegacyRef<HTMLDivElement>] = [{ isOver: false, canDrop: false }, undefined];

export const FlowElementWrapper: React.FC<{
    elementTypeClassName: string;
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode;
    editMenu?: (() => React.ReactNode) | undefined;
    onLinkDrop?: onLinkDropCallback | undefined;
    onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
}> = ({ elementTypeClassName, referenceElement, editMenu, onLinkDrop, onClick, children }) => {
    const [{ isOver, canDrop }, drop] = onLinkDrop
        ? useDrop({
              accept: DraggableType.LINK,
              canDrop: isDropValid(referenceElement, onLinkDrop),
              drop: (dragContext: DraggedLinkContext) => onLinkDrop(referenceElement, dragContext),
              collect: (monitor) => ({
                  isOver: !!monitor.isOver(),
                  canDrop: !!monitor.canDrop()
              })
          })
        : disabledDropping;
    return (
        <>
            <div className={`stroke-horizontal arrow${isOver && canDrop ? " can-drop" : ""}`} />
            <div className={`flow-element-wrapper${isOver ? (canDrop ? " can-drop" : " cannot-drop") : ""}`} ref={drop}>
                <div className={`flow-element ${elementTypeClassName}${editMenu ? " selected" : ""}`} onClick={onClick}>
                    {children}
                </div>
                {editMenu && editMenu()}
            </div>
        </>
    );
};
