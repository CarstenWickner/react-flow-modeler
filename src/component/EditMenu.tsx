import * as React from "react";

import { EditMenuItem } from "./EditMenuItem";
import { FlowElementReference } from "../model/FlowElement";
import { addContentElement, addDivergingGateway } from "../model/action/addElement";
import { addBranch } from "../model/action/addBranch";
import { isChangeNextElementAllowed } from "../model/action/changeNextElement";
import { removeElement, isRemoveElementAllowed } from "../model/action/removeElement";

import { SelectableElementType, EditActionResult, DraggableType } from "../types/EditAction";
import { FlowModelerProps, MenuOptions } from "../types/FlowModelerProps";
import { ElementType } from "../types/GridCellData";

const onClickStopPropagation = (event: React.MouseEvent): void => event.stopPropagation();

export class EditMenu extends React.Component<{
    targetType: SelectableElementType;
    referenceElement?: FlowElementReference;
    branchIndex?: number;
    menuOptions?: FlowModelerProps["options"]["editActions"];
    onChange: (change: (originalFlow: FlowModelerProps["flow"]) => EditActionResult) => void;
}> {
    isNextElementReferencedByOthers = (): boolean => {
        const { referenceElement, branchIndex } = this.props;
        if (!referenceElement) {
            return false;
        }
        return referenceElement.getFollowingElements()[branchIndex || 0].getPrecedingElements().length > 1;
    };

    renderMenuItem(
        options: MenuOptions,
        defaultClassName: string,
        onClick?: (event: React.MouseEvent) => void,
        dragType?: DraggableType
    ): React.ReactNode {
        const { targetType, referenceElement, branchIndex } = this.props;
        if (options && options.isActionAllowed && !options.isActionAllowed(targetType, referenceElement, branchIndex)) {
            return null;
        }
        const dragItem =
            dragType === DraggableType.LINK &&
            (targetType === ElementType.Content || targetType == ElementType.ConnectGatewayToElement) &&
            this.isNextElementReferencedByOthers()
                ? { type: dragType, originType: targetType, originElement: referenceElement, originBranchIndex: branchIndex }
                : undefined;
        return <EditMenuItem key={defaultClassName} options={options} defaultClassName={defaultClassName} onClick={onClick} dragItem={dragItem} />;
    }

    onAddContentElementClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addContentElement(originalFlow, targetType, {}, referenceElement, branchIndex));
        }
    };

    renderAddContentElementItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(menuOptions ? menuOptions.addFollowingContentElement : undefined, "add-content", this.onAddContentElementClick);
    }

    onAddDivergingGatewayClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addDivergingGateway(originalFlow, targetType, {}, referenceElement, branchIndex));
        }
    };

    renderAddDivergingGatewayItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(
            menuOptions ? menuOptions.addFollowingDivergingGateway : undefined,
            "add-gateway",
            this.onAddDivergingGatewayClick
        );
    }

    onAddDivergingBranchClick = (): void => {
        const { targetType, onChange, referenceElement } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            onChange((originalFlow) => addBranch(originalFlow, {}, referenceElement));
        }
    };

    renderAddDivergingBranchItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(menuOptions ? menuOptions.addDivergingBranch : undefined, "add-branch", this.onAddDivergingBranchClick);
    }

    renderChangeNextElementItem(): React.ReactNode {
        const { targetType, referenceElement, branchIndex, menuOptions } = this.props;
        if (isChangeNextElementAllowed(targetType, referenceElement, branchIndex)) {
            return this.renderMenuItem(menuOptions ? menuOptions.changeNextElement : undefined, "change-next", undefined, DraggableType.LINK);
        }
        return null;
    }

    onRemoveClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType === ElementType.Content || targetType === ElementType.ConnectGatewayToElement) {
            onChange((originalFlow) => removeElement(originalFlow, targetType, referenceElement, branchIndex));
        }
    };

    renderRemoveItem(): React.ReactNode {
        const { targetType, referenceElement, branchIndex, menuOptions } = this.props;
        if (isRemoveElementAllowed(targetType, referenceElement, branchIndex)) {
            return this.renderMenuItem(menuOptions ? menuOptions.removeElement : undefined, "remove", this.onRemoveClick);
        }
        return null;
    }

    render(): React.ReactNode {
        const menuItems = [
            this.renderAddContentElementItem(),
            this.renderAddDivergingGatewayItem(),
            this.renderAddDivergingBranchItem(),
            this.renderChangeNextElementItem(),
            this.renderRemoveItem()
        ];
        if (menuItems.some((item) => item)) {
            return (
                <div className="menu" onClick={onClickStopPropagation}>
                    {menuItems}
                </div>
            );
        }
        return null;
    }
}
