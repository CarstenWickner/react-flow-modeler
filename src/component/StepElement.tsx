import * as React from "react";

import { FlowElementWrapper } from "./FlowElementWrapper";

import { StepNode } from "../types/ModelElement";
import { onLinkDropCallback } from "../types/EditAction";

export class StepElement extends React.Component<{
    referenceElement: StepNode;
    editMenu: (() => React.ReactNode) | undefined;
    onLinkDrop: onLinkDropCallback | undefined;
    onSelect: (element: StepNode) => void;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, referenceElement } = this.props;
        onSelect(referenceElement);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { referenceElement, editMenu, onLinkDrop: onDrop, children } = this.props;
        return (
            <>
                <FlowElementWrapper
                    elementTypeClassName="step-element"
                    referenceElement={referenceElement}
                    editMenu={editMenu}
                    onLinkDrop={onDrop}
                    onClick={this.onClick}
                >
                    {children}
                </FlowElementWrapper>
                <div className="stroke-horizontal optional" />
            </>
        );
    }
}
