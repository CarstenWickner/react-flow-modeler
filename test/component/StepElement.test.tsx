import * as React from "react";
import { mount, shallow } from "enzyme";

import { StepElement } from "../../src/component/StepElement";
import { StepNode, ElementType } from "../../src/types/ModelElement";

const mockFlowElementReference = (id: string): StepNode => ({
    type: ElementType.StepNode,
    id,
    precedingElement: undefined,
    followingElement: undefined,
    columnIndex: 2,
    rowCount: 1
});

describe("renders correctly", () => {
    const onSelect = (): void => {};
    it("with minimal props", () => {
        const component = shallow(
            <StepElement
                referenceElement={mockFlowElementReference("element-id")}
                editMenu={(): React.ReactNode => <div className="edit-menu-placeholder" />}
                onLinkDrop={undefined}
                onSelect={onSelect}
            >
                {"text"}
            </StepElement>
        );
        expect(component).toMatchSnapshot();
    });
    it("when not selected", () => {
        const component = mount(
            <StepElement referenceElement={mockFlowElementReference("element-id")} editMenu={undefined} onLinkDrop={undefined} onSelect={onSelect}>
                {"text"}
            </StepElement>
        );
        const stepElement = component.find(".step-element");
        expect(stepElement.hasClass("selected")).toBe(false);
        expect(component.find(".edit-menu-placeholder").exists()).toBe(false);
    });
});
describe("calls onSelect", () => {
    const onSelect = jest.fn(() => {});
    const event = ({ stopPropagation: jest.fn(() => {}) } as unknown) as React.MouseEvent;
    it("on click event", () => {
        const stepNode = mockFlowElementReference("element-id");
        const component = mount(
            <StepElement referenceElement={stepNode} editMenu={undefined} onLinkDrop={undefined} onSelect={onSelect}>
                {"text"}
            </StepElement>
        );
        component.find(".step-element").prop("onClick")(event);
        expect(onSelect.mock.calls).toHaveLength(1);
        expect(onSelect.mock.calls[0]).toHaveLength(1);
        expect(onSelect.mock.calls[0][0]).toBe(stepNode);
    });
});
