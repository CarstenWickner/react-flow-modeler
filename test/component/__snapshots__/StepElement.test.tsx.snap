// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal props 1`] = `
<Fragment>
  <Component
    editMenu={[Function]}
    elementTypeClassName="step-element"
    onClick={[Function]}
    referenceElement={StepNode { "id": "element-id" }}
  >
    text
  </Component>
  <div
    className="stroke-horizontal optional"
  />
</Fragment>
`;
